c++/src/kj/compat/http.c++

```diff
@@ -4124,10 +4124,15 @@ public:
                                    leftoverBackingBuffer(kj::mv(leftoverBackingBuffer)),
                                    leftover(leftover) {}
 
+  // AsyncIoStream
   void shutdownWrite() override {
     stream->shutdownWrite();
   }
 
+  kj::Maybe<int> getFd() const override {
+    return stream->getFd();
```

Comment:
I'm not sure if it makes sense to implement `getFd()` for this class, since it's wrapping not just another stream, but a buffer as well. I guess it depends on what sort of semantics we want to provide `getFd()`. If we want to be able to guarantee that the returned FD will produce the same bytes that reading from this AsyncIoStream implementation would, then that might be a problem.

Reply:
In practice (in the usages of AsyncIoStreamWithInitialBuffer that I can see) I think the data in the leftover backing buffer has always previously been read from the `stream`'s FD, but it's true that the design of this class does not guarantee this.

Based on the documentation of `AsyncIoStream::getFd` which is:

>  // Get the underlying Unix file descriptor, if any. Returns nullptr if this object actually
>  // isn't wrapping a file descriptor.

I definitely wouldn't expect this to *always* return empty (which it does if we don't implement this).

An alternative could be to only return a non-empty FD if the `leftoverBackingBuffer` has been fully read/emptied. That would satisfy your criterion but only from the point at which `getFd()` starts returning non-empty. I'd be happy enough with that as a compromise since it would then at least sometimes be useful from a debugging perspective.

Reply:
These stream classes are all actually internal to `http.c++` so we could make it more explicit that any backing buffer provided should satisfy your proposed criterion (as all usages currently seemingly do).

Reply:
I'd be curious what @kentonv thinks.

Reply:
I agree with @harrishancock. The contract of `getFd()` is that if it returns an FD, then the stream itself does not implement any special behavior on top of that FD. For example, read()ing directly from the FD should produce identical results to read()ing from the stream.

That is not the case here.

In practice, `AsyncFdStream` (in `async-io.c++`) is probably the only stream that should ever implement `getFd()` to return non-null.